File: Leetcode/0006-zigzagconversation.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        res = [""] * numRows
        rows = numRows - 1
        down = True
        count = 0
        for letter in s:
            if down:
                res[count] += letter
                count += 1
            if not down:
                count -= 1
                res[count] += letter
            if count <= 0:
                down = True
                count += 1
            if count > rows:
                down = False
                count -= 1
        return "".join(res)

File: Leetcode/0006-zigzagconversation.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        int n = s.size(), wavelength = (numRows-1) << 1; 
        if(n <= numRows || numRows == 1) return s;
        string ans = "";
        for(int i = 0; i < numRows; i++){
            if(i == 0 || i == numRows - 1){
                int idx = i;
                while(idx < n){
                    ans += s[idx];
                    idx+=wavelength;
                }
            }else{
                int diff = wavelength-(i<<1);
                int idx = i;
                while(idx < n){
                    ans+=s[idx];
                    idx += diff;
                    diff = wavelength-diff;
                }
            }
        }
        return ans;
    }
};